pilot/pkg/secrets/kube/secrets.go

```diff
@@ -0,0 +1,135 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kube
+
+import (
+	"strings"
+
+	v1 "k8s.io/api/core/v1"
+	informersv1 "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/tools/cache"
+
+	"istio.io/istio/pilot/pkg/secrets"
+	"istio.io/pkg/log"
+)
+
+const (
+	// The ID/name for the certificate chain in kubernetes generic secret.
+	GenericScrtCert = "cert"
+	// The ID/name for the private key in kubernetes generic secret.
+	GenericScrtKey = "key"
+	// The ID/name for the CA certificate in kubernetes generic secret.
+	GenericScrtCaCert = "cacert"
+
+	// The ID/name for the certificate chain in kubernetes tls secret.
+	TLSSecretCert = "tls.crt"
+	// The ID/name for the k8sKey in kubernetes tls secret.
+	TLSSecretKey = "tls.key"
+	// The ID/name for the CA certificate in kubernetes tls secret
+	TLSSecretCaCert = "ca.crt"
+
+	// GatewaySdsCaSuffix is the suffix of the sds resource name for root CA. All resource
+	// names for gateway root certs end with "-cacert".
+	GatewaySdsCaSuffix = "-cacert"
+)
+
+type SecretsController struct {
+	secrets informersv1.SecretInformer
+}
+
+var _ secrets.Controller = &SecretsController{}
+
+func NewSecretsController(informer informersv1.SecretInformer) *SecretsController {
+	return &SecretsController{informer}
+}
+
+func (s *SecretsController) GetKeyAndCert(name, namespace string) (key []byte, cert []byte) {
+	k8sSecret, err := s.secrets.Lister().Secrets(namespace).Get(name)
+	if err != nil {
+		return nil, nil
+	}
+
+	return extractKeyAndCert(k8sSecret)
+}
+
+func (s *SecretsController) GetCaCert(name, namespace string) (cert []byte) {
+	strippedName := strings.TrimSuffix(name, GatewaySdsCaSuffix)
+	k8sSecret, err := s.secrets.Lister().Secrets(namespace).Get(strippedName)
+	var rootCert []byte
+	if err != nil {
+		// Could not fetch cert, look for legacy secret with -cacert suffix
+		k8sSecret, caCertErr := s.secrets.Lister().Secrets(namespace).Get(name)
+		if caCertErr != nil {
+			return nil
+		}
+		rootCert = extractRoot(k8sSecret)
+	} else {
+		rootCert = extractRoot(k8sSecret)
+		// Secret exists, but does not have the ca cert. Fall back to -cacert secret
+		if rootCert == nil {
+			k8sSecret, caCertErr := s.secrets.Lister().Secrets(namespace).Get(name)
+			if caCertErr != nil {
+				return nil
+			}
+			rootCert = extractRoot(k8sSecret)
+		}
+	}
+	return rootCert
+}
+
+// extractKeyAndCert extracts server key, certificate
+func extractKeyAndCert(scrt *v1.Secret) (key, cert []byte) {
+	if len(scrt.Data[GenericScrtCert]) > 0 {
+		cert = scrt.Data[GenericScrtCert]
+		key = scrt.Data[GenericScrtKey]
+	} else {
+		cert = scrt.Data[TLSSecretCert]
```

Comment:
do we also need to check the existence of TLSSecretCert?

Reply:
Good point - maybe. Let me write some unit tests to answer that